### PR TITLE
posix: device_io: implement fdopen(), fileno(), pread(), pselect(), pwrite()

### LINKED ISSUE #24966: sockets/posix: poll()/select() implementation should be factored out from sockets subsys into POSIX subsys
POSIX poll() was initially implemented in the scope of BSD Sockets API, and currently located in subsys/net/lib/sockets/ . It is however a generic POSIX facility, and was generalized to apply to arbitrary file descriptors (and objects represented by them). As we grow implementations of such objects (eventfd, unix domain sockets), dependency on networking sockets becomes a problem, as requires enabling unrelated config options and leads to pulling more code into user applications.

To address this, poll() (and by extension, select()) implementation should be migrated to lib/posix/.

### LINKED ISSUE #66932: posix: device_io: implement `fdopen()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`fdopen()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/fdopen.html

### LINKED ISSUE #66938: posix: device_io: implement `fileno()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`fileno()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/fileno.html

### LINKED ISSUE #66946: posix: device_io: implement `pread()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`pread()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/pread.html

### LINKED ISSUE #66947: posix: device_io: implement `pselect()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`pselect()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/pselect.html

### LINKED ISSUE #66948: posix: device_io: implement `pwrite()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`pwrite()` is required as part of `POSIX_DEVICE_IO` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/pwrite.html

### PR DESCRIPTION
**Pull request by cfriedt**
Implement the remaining functions in the `POSIX_DEVICE_IO` Option Group.

This Option Group is requried for the PSE51, PSE52, PSE53, and PSE54 subprofiles and likely others.

[Doc Preview](https://builds.zephyrproject.io/zephyr/pr/73978/docs/services/portability/posix/index.html)

Fixes #24966
Fixes #66932 
Fixes #66938 
Fixes #66946 
Fixes #66947 
Fixes #66948

### CODE REVIEW

**Comment by cfriedt** [comment]
This one has a false positive regarding the reserved word "remove" due to the `remove()` function existing in a file that was touched.

No changes made there, although it is an ISO C function, and not a POSIX one. That would be a good bugfix at some point.

**Comment by cfriedt** [comment]
Fixes for a few bugs prepended.

**Review by SeppoTakalo — DISMISSED** [review]

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

File: lib/posix/options/Kconfig.device_io
Comment (jukkar) [inline lib/posix/options/Kconfig.device_io:50]: The help text is slightly misleading as user is not able to y this option because it is hidden.

Comment (cfriedt) [reply lib/posix/options/Kconfig.device_io:50]: Will fix after the TSC meeting tomorrow.

**Review by ycsin — DISMISSED** [review]

File: lib/os/fdtable.c
Comment (ycsin) [inline lib/os/fdtable.c:77]: I guess we might need to add `FILE` to https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/checkpatch/typedefsfile

File: subsys/net/lib/ptp/clock.c
Comment (cfriedt) [reply subsys/net/lib/ptp/clock.c:58]: Yes, there is a separate PR for it, but I needed to have it here too because I found that otherwise there would be a compile error.

File: lib/os/fdtable.c
Comment (cfriedt) [reply lib/os/fdtable.c:77]: Maybe - I thought it might have been because the minimal libc defines `stdin`, `stdout`, `stderr` as macros. Hard to say.

Comment (ycsin) [reply lib/os/fdtable.c:77]: It seems to fix the issue in vscode locally, created #74507

**Comment by cfriedt** [comment]
temporary DNM until merge conflicts and all comments have been addressed.

File: lib/posix/options/Kconfig.device_io
Comment (cfriedt) [reply lib/posix/options/Kconfig.device_io]: Should be fixed now. I applied the same pattern to other non-user-selectable Kconfig as well.

**Comment by cfriedt** [comment]
* rebased
* fixed merge conflicts
* updated language for non-user-selectable Kconfig options (i.e. do not use "Select 'y'" if the user cannot select y)

**Comment by cfriedt** [comment]
Other than the false positive in coding guidelines, this should probably be all green. Please revisit when you have a moment @ycsin, @jukkar, @SeppoTakalo

**Review by ycsin — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**
